src/topology/tactic.lean

```diff
@@ -0,0 +1,77 @@
+/-
+Copyright (c) 2020 Reid Barton. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Reid Barton
+-/
+import tactic.auto_cases
+import tactic.tidy
+import tactic.with_local_reducibility
+import topology.basic
+
+/-!
+# `continuity` tactic
+
+Automatically solve goals of the form `continuous f`.
+
+Mark lemmas with `@[continuity]` to add them to the set of lemmas
+used by `continuity`. Note: `to_additive` doesn't know yet how to
+copy the attribute to the additive version.
+-/
+
+/-- User attribute used to mark tactics used by `continuity`. -/
+@[user_attribute]
+meta def continuity : user_attribute :=
+{ name := `continuity,
+  descr := "lemmas usable to prove continuity" }
+
+-- Mark some continuity lemmas already defined in `topology.basic`
+attribute [continuity]
+  continuous_id
+  continuous_const
+
+namespace tactic
+
+/--
+Tactic to apply `continuous.comp` when appropriate.
+
+Applying `continuous.comp` is not always a good idea, so we have some
+extra logic here to try to avoid bad cases.
+
+* If the function we're trying to prove continuous is actually
+  constant, and that constant is a function application `f z`, then
+  continuous.comp would produce new goals `continuous f`, `continuous
+  (λ _, z)`, which is silly. We avoid this by failing if we could
+  apply continuous_const.
+
+* continuous.comp will always succeed on `continuous (λ x, f x)` and
+  produce new goals `continuous (λ x, x)`, `continuous f`. We detect
+  this by failing if a new goal can be closed by applying
+  continuous_id.
+-/
+meta def apply_continuous.comp : tactic unit :=
+`[fail_if_success { exact continuous_const };
+  refine continuous.comp _ _;
+  fail_if_success { exact continuous_id }]
+
+/-- List of tactics used by `continuity` internally. -/
+meta def continuity_tactics : list (tactic string) :=
+[
+  `[apply_rules continuity]            >> pure "apply_rules continuity",
+  auto_cases,
```

Comment:
What is `auto_cases` doing for us?

Reply:
Does anything break if this is removed?

Reply:
It seems to work fine without. @rwbarton? Presumably there's a use case in your homotopy theory repository?